### PR TITLE
Customize shebang 739

### LINKED ISSUE #739: The shebang line at the beginning of the job script should be customizable
This line is currently hardcoded to `#!/bin/bash`, but it should be customizable in some way.

Some use cases:
- appending `-l` to the end (necessary for instance on Piz Daint in order to enable the `module` command - see https://user.cscs.ch/getting_started/running_jobs)
- using a bash executable at a different location
- using a different shell (if we are supporting anything else than bash)

https://github.com/aiidateam/aiida_core/blob/8f52eb0cea65552e423c0a6de9568c2b5a7c120c/aiida/scheduler/__init__.py#L148

### PR DESCRIPTION
**Pull request by lekah**
This pull request resolves #739 by enforcing the customization of the shebang line for each computer.
For existing computers, the previous default is taken (#!/bin/bash)

### CODE REVIEW

File: aiida/backends/sqlalchemy/models/computer.py [hunk 1]
```diff
@@ -98,6 +98,13 @@ def get_workdir(self):
             raise ConfigurationError('No workdir found for DbComputer {} '.format(
                 self.name))
 
+    def get_shebang(self):
+        try:
+            return self._metadata['shebang']
+        except KeyError:
+            raise ConfigurationError('No workdir found for DbComputer {} '.format(
```
Comment (sphuber) [inline aiida/backends/sqlalchemy/models/computer.py]: I think you forgot to change the exception message

File: aiida/scheduler/__init__.py [hunk 2]
```diff
@@ -145,11 +145,10 @@ def get_submit_script(self, job_tmpl):
 
         empty_line = ""
 
-        shebang = "#!/bin/bash"
 
         # I fill the list with the lines, and finally join them and return
         script_lines = []
-        script_lines.append(shebang)
+        script_lines.append(job_tmpl.shebang)
```
Comment (giovannipizzi) [inline aiida/scheduler/__init__.py]: Can we put here a default? Something like: `if job_tmpl.shebang is None: ... else: ...`? I see you changed all plugins, which is good, but I think there are a few that are developed as external plugins and might not be aware of this change.

Comment (giovannipizzi) [reply aiida/scheduler/__init__.py]: (maybe verifying that the default is None, and leaving the option for an empty string to be dealt differently, if a user does not want the shebang line at all? I don't know if there is ever this need, but one never knows). Something like `elif job_tmpl == "": #dont print anything, not even an empty line`

File: aiida/scheduler/plugins/test_pbspro.py [hunk 3]
```diff
@@ -908,6 +908,7 @@ def test_submit_script(self):
         s = PbsproScheduler()
 
         job_tmpl = JobTemplate()
+        job_tmpl.shebang = '#!/bin/bash -l'
```
Comment (giovannipizzi) [inline aiida/scheduler/plugins/test_pbspro.py]: You put a `-l` as a way to verify the current implementation?
Maybe if you implement also the case with `shebang` not specified, you can have a copy of one of these tests to verify that even if you don't set it, you get the default value?

File: aiida/orm/implementation/sqlalchemy/computer.py [hunk 4]
```diff
@@ -232,8 +233,20 @@ def set_workdir(self, val):
         metadata = self._get_metadata()
         metadata['workdir'] = val
         self._set_metadata(metadata)
-        #else:
-        #    raise ModificationNotAllowed("Cannot set a property after having stored the entry")
+
+    def get_shebang(self):
```
[the remark this answers is not in the document]
Comment (lekah) [reply aiida/orm/implementation/sqlalchemy/computer.py]: Don't think it should. The fact that there is a member called dbcomputer is a choice of the implementation. Some other backend could call it _dbcomputer or whatever, it's not enforced. As well as the function get_shebang, also not enforced. I would leave it where it is.

File: aiida/orm/implementation/general/computer.py [hunk 5]
```diff
@@ -712,9 +712,9 @@ def set_shebang(self, val):
         :param str val: A valid shebang line
         """
         if not isinstance(val, basestring):
-            raise ValueError("Input has to be a string")
+            raise ValueError("{} is invalid. Input has to be a string".format(val))
         if not val.startswith('!#'):
-            raise ValueError("A shebang line has to start with #!")
+            raise ValueError("{} is invalid. A shebang line has to start with #!".format(val))
```
Comment (sphuber) [inline aiida/orm/implementation/general/computer.py]: Not sure if this was intentional irony, but you actually swapped the `#` and `!` around in the conditional :smile:

Comment (giovannipizzi) [reply aiida/orm/implementation/general/computer.py]: I think that if this is fixed, we still need @sphuber to accept the changes before merging?

**Pull request merged**